SCIM Workshop Blog Post (#1408)
oktadev/okta-blog: _source/_posts/2023-07-28-scim-workshop.md

Hello SaaS developers! Do you currently have a need to provision, de-provision, and/or sync users across services? If so, then SCIM is the right tool for you! What is SCIM, you ask? SCIM stands for System for Cross-domain Identity Management. And it is an open standard protocol that allows us to manage identities across systems as well as manage common user lifecycles. So in this workshop, our goal is to build a SCIM server in a SaaS app using this standard protocol and, in the end, connect it to an identity provider such as Okta. 

In this workshop, we'll cover the following:

{% include toc.md %}

In this workshop, you will learn how to implement SCIM and support provisioning from multiple organizations/tenants. Although we'll integrate with Okta as an example, remember that almost every identity provider on the web has SCIM support! 

# Understanding SCIM 

## What problems does SCIM solve?
There are many challenges with managing user identities across services:
- SaaS developers want to know as soon as someone joins a company so they can set up an account 
- SaaS developers also want to know when someone leaves a company to deactivate their account and associated resources.

> - Manually managing users through CSVs

Author: ```suggestion
```

- Manually adding accounts at scale

SCIM, the System for Cross-domain Identity Management, is an open standard which addresses those problems. SCIM: 
- Provides near-instant updates to the SaaS app whenever someone joins, moves inside of, or leaves a company

> - Allows management of user identities in cloud-based applications and services - i.e., you will have awareness of your users' up-to-date information

Author: ```suggestion
- Allows management of user identities in cloud-based applications and services - i.e., enterprise customers expect their SaaS apps to immediately and effortlessly reflect user onboarding, offboarding, and other changes. 
```

```
import { scimRoute } from './scim';
app.use('/scim/v2', scimRoute);
```

Note: `/scim/v2` will append to every SCIM route. You can change this static path as needed, as there isn't a specific URL path explicitly mentioned by the SCIM spec. However, [Okta's SCIM Docs](https://developer.okta.com/docs/guides/scim-provisioning-integration-prepare/main/#base-url) recommend using `/scim/v2` unless unusual factors in your environment necessitate a different URL. 

## Build the SCIM interface:

We'll need to build each CRUD endpoint required by the SCIM spec and format the responses in JSON. After that, we'll test our functions with Postman to see that they provide the responses required by the spec. 

We'll need to use express and Prisma in `scim.ts`, so import them:

```
import express from 'express';
import { Prisma, PrismaClient } from '@prisma/client';
```

Some SCIM endpoints in `scim.ts` will return user information from the Todo app's database. To easily retrieve this information from the SCIM endpoints, implement an `IUserSchema` interface and instantiate it as the defaultUserSchema. The information in the `IUserSchema` matches the SCIM spec, as you'll see soon.  

> 

Author: ```suggestion

When supporting a single organization, we can assign it the ID 1 and hardcode that value. However, as the app scales to integrate with multiple customers' SCIM servers, this constant would be replaced with a function to look up the appropriate org ID based on the context of the request.

```
